server/src/main/java/org/elasticsearch/bootstrap/Spawner.java

```diff
@@ -75,16 +87,39 @@ void spawnNativeControllers(final Environment environment, final boolean inherit
                 );
                 throw new IllegalArgumentException(message);
             }
-            final Process process = spawnNativeController(spawnPath, environment.tmpFile(), inheritIo);
+            final Process process = spawnNativeController(spawnPath, environment.tmpFile());
+            // The process _shouldn't_ write any output via its stdout or stderr, but if it does then
+            // it will block if nothing is reading that output. To avoid this we can pipe the
+            // outputs and create pump threads to write any messages there to the ES log.
+            startPumpThread(info.getName(), "stdout", process.getInputStream());
+            startPumpThread(info.getName(), "stderr", process.getErrorStream());
             processes.add(process);
         }
     }
 
+    private void startPumpThread(String componentName, String streamName, InputStream stream) {
+        String loggerName = componentName + "-controller-" + streamName;
+        final Logger logger = LogManager.getLogger(loggerName);
+        Thread t = new Thread(() -> {
+            try (var br = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    // since we do not expect native controllers to ever write to stdout/stderr, we always log at warn level
+                    logger.warn(line);
+                }
+            } catch (IOException e) {
+                logger.error("error while reading " + streamName, e);
```

Comment:
```suggestion
                logger.error("error while reading " + loggerName, e);
```

Otherwise just "error while reading stderr" might be a bit unclear about which stderr to the average person reading the log file.

Reply:
The logger name itself is descriptive. I only added the message at all because there is no error method which takes only a throwable.